hal/src/thumbv7em/icm.rs

```diff
@@ -0,0 +1,1612 @@
+    #[inline]
+    fn reset_rcfg_to_default(&mut self) {
+        self.rcfg = Rcfg::default();
+    }
+    /// Set [`Rnext`]
+    #[inline]
+    fn set_rnext(&mut self, next: Rnext) {
+        self.rnext = next;
+    }
+}
+
+impl<N: IcmRegionNum> IcmRegionDesc<N> {
+    /// The length of data for the ICM engine to transfer,
+    /// expressed as number of `blocks - 1`.
+    #[inline]
+    pub fn set_rctrl(mut self, ctrl: Rctrl) {
+        self.rctrl = ctrl;
+    }
+}
+
+/// Structure ICM Secondary Region Descriptor area.
+///
+/// Follows C-structure conventions and is 16-byte aligned
+///
+/// Used to build the linked lists for non-contiguous memory
+#[repr(C)]
+#[repr(align(16))]
+pub struct IcmSecondaryRegionDesc {
+    /// the first byte address of the Region.
+    pub raddr: Raddr,
+    /// Configuration Structure Member.
+    pub rcfg: Rcfg,
+    /// Not used in Secondary Region Descriptor
+    _pad: u32,
+    /// Next Address Structure Member.
+    pub rnext: Rnext,
+}
+
+impl IcmRegionDescT for IcmSecondaryRegionDesc {
+    fn set_raddr<T>(&mut self, addr: *const T) {
+        self.raddr.set_raddr(addr);
+    }
+    /// Set [`Rcfg`]
+    #[inline]
+    fn set_rcfg(&mut self, cfg: Rcfg) {
+        self.rcfg = cfg;
+    }
+    /// Reset [`Rcfg`] to default values
+    #[inline]
+    fn reset_rcfg_to_default(&mut self) {
+        self.rcfg = Rcfg::default();
+    }
+    /// Set [`Rnext`]
+    #[inline]
+    fn set_rnext(&mut self, next: Rnext) {
+        self.rnext = next;
+    }
+}
+
+impl IcmSecondaryRegionDesc {
+    pub const fn default() -> Self {
+        IcmSecondaryRegionDesc {
+            raddr: Raddr::default(),
+            rcfg: Rcfg::default(),
+            _pad: 0,
+            rnext: Rnext::default(),
+        }
+    }
+}
+
+/// ICM Hash Area
+///
+/// Follows C-structure conventions and is 128-byte aligned
+///
+/// >**Important**
+/// >
+/// >ICM engine will **read** or **write** to wherever this
+/// >is instantiated in memory, based on the [`Icm::set_hash_addr()`]
+/// >so the user must ensure that this variable lives long enough or is
+/// >static
+#[derive(Debug)]
+#[repr(C)]
+#[repr(align(128))]
+pub struct IcmHashArea {
+    pub region0: [u32; 8],
+    pub region1: [u32; 8],
+    pub region2: [u32; 8],
+    pub region3: [u32; 8],
+}
+
+impl IcmHashArea {
+    pub fn region0(&self) -> [u32; 8] {
+        self.region0
+    }
+    pub fn region1(&self) -> [u32; 8] {
+        self.region1
+    }
+    pub fn region2(&self) -> [u32; 8] {
+        self.region2
+    }
+    pub fn region3(&self) -> [u32; 8] {
+        self.region3
+    }
+    pub const fn default() -> Self {
+        IcmHashArea {
+            region0: [0; 8],
+            region1: [0; 8],
+            region2: [0; 8],
+            region3: [0; 8],
+        }
+    }
+}
+
+/// Region Start Address Structure
+///
+/// Follows C-structure conventions
+#[derive(Clone, Copy)]
+#[repr(C)]
+pub struct Raddr {
+    pub raddr: u32,
+}
+
```

Comment:
I think you're matching the datasheet here, but I don't like `Raddr` as a name. It's very terse. Without a lot of context, I wouldn't know what it represents.

That being said, if you rename it here, you probably ought to be more verbose throughout the module, e.g. in function names. Personally, I would probably prefer `RegionAddress`, `get_region_address`, etc. But I can see counter arguments.

Reply:
Adopted `RegionAddress`, similarly for `Rctrl` etc.